src/axom/quest/MarchingCubes.hpp

```diff
@@ -0,0 +1,283 @@
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/*!
+ * \file MarchingCubes.hpp
+ *
+ * \brief Consists of class implementing marching cubes algorithm to
+ * compute iso-surface from a scalar field in a blueprint mesh.
+ */
+
+#ifndef AXOM_PRIMAL_MARCHINGCUBES_H_
+#define AXOM_PRIMAL_MARCHINGCUBES_H_
+
+// Axom includes
+#include "axom/config.hpp"
+#include "axom/mint/mesh/Mesh.hpp"
+
+// Conduit includes
+#include "conduit_node.hpp"
+
+// C++ includes
+#include <string>
+
+namespace axom
+{
+namespace quest
+{
+/*!
+ * \@brief Class implementing marching cubes algorithm on a single
+ * structured mesh domain within a multi-domain mesh.
+ *
+ * Implementation is for 2D (marching squares) and 3D (marching cubes).
+ *
+ * The input mesh is a Conduit::Node following the Mesh Blueprint
+ * convention.
+ *
+ * \sa MarchingCubes
+ */
+class MarchingCubesSingleDomain
+{
+public:
+  /*!
+   * \brief Constructor for applying algorithm in multi-level mesh context.
+   *
+   * \param [in] dom Blueprint single-domain mesh containing scalar field.
+   * \param [in] coordsetName Name of blueprint point coordinate set.
+   * \param [in] maskField Cell-based std::int32_t mask field.  If provided,
+   *             cells where this field evaluates to false are skipped.
+   *
+   * Some data from \a dom may be cached by the constructor.
+   * Any change to it after the constructor leads to undefined behavior.
+   * See set_domain(const conduit::Node &) for requirements on \a dom.
+   *
+   * The mesh coordinates should be contiguous.  See
+   * conduit::blueprint::is_contiguous().  In the future, this
+   * requirement may be relaxed, possibly at the cost of a
+   * transformation and storage of the temporary contiguous layout.
+   */
+  MarchingCubesSingleDomain(const conduit::Node &dom,
+                            const std::string &coordsetName,
+                            const std::string &maskfield);
+
+  //! @brief Spatial dimension of domain.
+  int dimension() const { return m_ndim; }
+
+  /*!
+    @brief Set the field containing the nodal function.
+    \param [in] fcnField Name of node-based scalar function values.
+  */
+  void set_function_field(const std::string &fcnField);
+
+  /*!
+    @brief Set the output surface mesh object.
+  */
+  void set_output_mesh(axom::mint::Mesh *surfaceMesh);
+
+  /*!
+    @brief Save the originating cell index in output mesh.
+
+    For each contour mesh cell generated, the 1D index of the
+    generating computational mesh cell is set in the \a cellIdField of
+    the contour mesh.  If the field doesn't exist, it will be created.
+    To disable, set \a outputCellIds to empty.
+  */
+  void set_cell_id_field(const std::string &cellIdField)
+  {
+    m_cellIdField = cellIdField;
+  }
+
+  /*!
+   * \brief Computes the iso-surface.
+   *
+   * \param [in] contourVal iso-contour value
+   *
+   * Compute iso-surface using the marching cubes algorithm.
```

Comment:
Please update the doxygen here or for the class to indicate which functions need to be called before `compute_iso_surface`.  Presumably, `set_output_mesh()` and `set_function_field()`, and (optionally) `set_cell_id_field`.

Would it make sense for the output mesh and function field (and other parameters) to be supplied when calling this function?

Reply:
It's possible to put those parameters in the `compute_iso_surace` interface, but I think it's a bit cleaner to not to, because they'd have to appear in the `MarchingCubesSingleDomain::compute_iso_surface` argument list as well.  Also, using `set_...()` methods was the convention in `DistributedClosestPoints`.